### PR TITLE
MM-46402: Use pointer to pointer to set pointer to nil

### PR DESCRIPTION
**Pull request by ashishbhate**
#### Summary

- We need a pointer to a pointer to set the original pointer to nil
- The original task was not being set to `nil` (the local variable containing the pointer was being set to nil). The cancel func was being called even though the task had already been cancelled. This repeated cancellation was causing a panic because we were trying to close a channel that had already been closed the first time the task was cancelled.

#### Ticket Link

- https://mattermost.atlassian.net/browse/MM-46402

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: app/server.go [hunk 1]
```diff
@@ -2060,12 +2060,12 @@ func withMut(mut *sync.Mutex, f func()) {
 	f()
 }
 
-func cancelTask(mut *sync.Mutex, task *model.ScheduledTask) {
+func cancelTask(mut *sync.Mutex, task **model.ScheduledTask) {
 	mut.Lock()
 	defer mut.Unlock()
-	if task != nil {
-		task.Cancel()
-		task = nil
```
Comment (ashishbhate) [inline app/server.go:2068]: Here just the local variable (`task`) containing the pointer was being set to `nil`. 
We needed to set the pointer of the global task variable (`a.ch.dndTask` or `a.ch.postReminderTask`, depending on where this function was called) to nil.

File: app/server_test.go [hunk 2]
```diff
@@ -565,3 +566,11 @@ func TestSentry(t *testing.T) {
 		}
 	})
 }
+
+func TestCancelTaskSetsTaskToNil(t *testing.T) {
+	var taskMut sync.Mutex
+	task := model.CreateRecurringTaskFromNextIntervalTime("a test task", func() {}, 5*time.Minute)
+	require.NotNil(t, task)
+	cancelTask(&taskMut, &task)
+	require.Nil(t, task)
```
Comment (agarciamontoro) [inline app/server_test.go:575]: Could we call cancelTask here again, which was the cause of the problem?

File: app/server.go [hunk 3]
```diff
@@ -2060,12 +2060,12 @@ func withMut(mut *sync.Mutex, f func()) {
 	f()
 }
 
-func cancelTask(mut *sync.Mutex, task *model.ScheduledTask) {
+func cancelTask(mut *sync.Mutex, task **model.ScheduledTask) {
```
Comment (agarciamontoro) [inline app/server.go:2063]: This may be confusing. Maybe rename it to taskPointer?

**Review by agarciamontoro — APPROVED** [review]
Aaaaah, this makes sense! Great catch, @ashishbhate, I completely missed this when first looking at the code. Thank you for the fix :)

I left a couple of optional comments below.

File: app/server_test.go (hunk 2 above)
Comment (ashishbhate) [reply app/server_test.go:575]: Done

File: app/server.go (hunk 3 above)
Comment (ashishbhate) [reply app/server.go]: Done

File: app/server.go [hunk 4]
```diff
@@ -2060,12 +2060,12 @@ func withMut(mut *sync.Mutex, f func()) {
...
 
-func cancelTask(mut *sync.Mutex, task *model.ScheduledTask) {
+func cancelTask(mut *sync.Mutex, taskPointer **model.ScheduledTask) {
 	mut.Lock()
 	defer mut.Unlock()
-	if task != nil {
-		task.Cancel()
-		task = nil
+	if *taskPointer != nil {
+		(**taskPointer).Cancel()
```
Comment (agnivade) [inline app/server.go:2067]: Why `**` here? This should work with a pointer method as well right?

Comment (ashishbhate) [reply app/server.go:2067]: Yep, a single `*` works as well.

**Review by agnivade — APPROVED** [review]
Nice!

**Comment by ashishbhate** [comment]
@agarciamontoro I'm having trouble connecting to our internal servers. If the report is passing, could you please merge this PR?

**Comment by agnivade** [comment]
https://community.mattermost.com/private-core/pl/xwygwzxgdpgybpfoprfjy6wpjh

You'll need to compare the failures against a recent master test run. Looking at the results, they look expected to me. Failures are unrelated.

**Pull request merged**